pkg/sources/forward/data_forward.go

```diff
@@ -311,16 +311,26 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
 	// publish source watermark and assign IsLate attribute based on new event time.
 	var writeMessages []*isb.WriteMessage
 	var transformedReadMessages []*isb.ReadMessage
+	latestEtMap := make(map[int32]int64)
+
 	for _, m := range transformerResults {
 		writeMessages = append(writeMessages, m.writeMessages...)
 		for _, message := range m.writeMessages {
 			// we convert each writeMessage to isb.ReadMessage by providing its parent ReadMessage's ReadOffset.
 			// since we use message event time instead of the watermark to determine and publish source watermarks,
 			// time.UnixMilli(-1) is assigned to the message watermark. transformedReadMessages are immediately
 			// used below for publishing source watermarks.
+			if latestEt, ok := latestEtMap[m.readMessage.ReadOffset.PartitionIdx()]; !ok || message.EventTime.UnixNano() < latestEt {
```

Comment:
For all of these debugging purpose logic, we should think about using an approach that does not affect the performance. e.g. turn it off by default.

Reply:
+1